src/protocols/secure_channel/PASESession.cpp

```diff
@@ -578,115 +651,131 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
+    size_t verifier_len = kMAX_Hash_Length;
 
     System::PacketBufferHandle resp;
 
-    Spake2pErrorType spake2pErr = Spake2pErrorType::kUnexpected;
-
-    uint16_t encryptionKeyId = 0;
+    uint16_t spake2pErr = kProtocolCodeInvalidParam;
 
     ChipLogDetail(SecureChannel, "Received spake2p msg2");
 
-    VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
-    VerifyOrExit(buf_len == sizeof(encryptionKeyId) + kMAX_Point_Length + kMAX_Hash_Length,
-                 err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
 
-    encryptionKeyId = chip::Encoding::LittleEndian::Read16(buf);
-    msg->ConsumeHead(sizeof(encryptionKeyId));
-    buf     = msg->Start();
-    buf_len = msg->DataLength();
+    const uint8_t * Y;
+    size_t Y_len = 0;
 
-    ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", encryptionKeyId);
-    SetPeerKeyId(encryptionKeyId);
+    const uint8_t * peer_verifier;
+    size_t peer_verifier_len = 0;
 
-    err = mSpake2p.ComputeRoundTwo(buf, kMAX_Point_Length, verifier, &verifier_len_raw);
-    SuccessOrExit(err);
-    VerifyOrExit(CanCastTo<uint16_t>(verifier_len_raw), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
-    verifier_len = static_cast<uint16_t>(verifier_len_raw);
+    uint32_t decodeTagIdSeq = 0;
 
-    {
-        const uint8_t * hash = &buf[kMAX_Point_Length];
-        err                  = mSpake2p.KeyConfirm(hash, kMAX_Hash_Length);
-        if (err != CHIP_NO_ERROR)
-        {
-            spake2pErr = Spake2pErrorType::kInvalidKeyConfirmation;
-            SuccessOrExit(err);
-        }
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-        err = mSpake2p.GetKeys(mKe, &mKeLen);
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    Y_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(Y));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    peer_verifier_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(peer_verifier));
+
+    SuccessOrExit(err = mSpake2p.ComputeRoundTwo(Y, Y_len, verifier, &verifier_len));
+
+    err = mSpake2p.KeyConfirm(peer_verifier, peer_verifier_len);
+    if (err != CHIP_NO_ERROR)
+    {
+        spake2pErr = kProtocolCodeNoSharedRoot;
         SuccessOrExit(err);
     }
 
+    SuccessOrExit(err = mSpake2p.GetKeys(mKe, &mKeLen));
+
     {
-        Encoding::PacketBufferWriter bbuf(System::PacketBufferHandle::New(verifier_len));
-        VerifyOrExit(!bbuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+        size_t data_len = EstimateTLVStructOverhead(verifier_len, 1);
```

Comment:
These `data_len` in each message handler to estimate size should be renamed to `max_msg_len` or something similar, since they actually do not ever reflect the true length of the message. Suggest making them `const` as well, to make clear that they are not reused